Enable PHPStan in tests

Majkl578 opened the pull request:
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no

#### Summary

Enabling PHPStan in tests/ at level 3, just like for lib/. :tada: 
Rewritten some Prophecy-based unit tests into PHPUnit mock system to avoid having another (unofficial) extension for Prophecy.

Replaces incomplete #3203 with no changes like (_Class_ -> _mixed_).

Currently depends on my fork of phpstan-phpunit due to https://github.com/phpstan/phpstan-phpunit/pull/28.

guilhermeblanco committed:
Decouple unique index from unique constraint

guilhermeblanco committed:
Added flag support to unique constraint

guilhermeblanco committed:
Fixed tests

guilhermeblanco committed:
Removed identifier generator type from schema table definition

guilhermeblanco committed:
Fixing rebase with master

guilhermeblanco committed:
Fixed undefined result of rebase

guilhermeblanco committed:
Moved getSequencePrefix() from ClassMetadata (ORM) to AbstractPlatform (DBAL)

morozov committed:
Updated UPGRADE.md for 3.x

morozov committed:
[BC] Removed support for PDO::PARAM_INPUT_OUTPUT in Statement::bindParam()

morozov committed:
[BC] Replaced PDO constants with values

morozov committed:
[BC] Replaced extension of \PDOStatement with a composition to avoid having to replicate the \PDOStatement interface in ResultStatement

Majkl578 committed:
Bump version to 3.0.x-dev

Majkl578 committed:
Drop Doctrine\DBAL\Version in favor of Ocramius\PackageVersions

BenMorel committed:
Drop Drizzle support

BenMorel committed:
Add note to UPGRADE.md

BenMorel committed:
Introduce a null SQL logger

BenMorel committed:
Make NullLogger final

BenMorel committed:
One-liner for Configuration::getSQLLogger()

BenMorel committed:
Change the SQLLogger interface signature

BenMorel committed:
Add UPGRADE note

BenMorel committed:
Ignore NullLogger in code coverage

BenMorel committed:
Make SQLLogger implementations final

BenMorel committed:
Update UPGRADE.md

BenMorel committed:
Do not accept null in Configuration::setSQLLogger()

BenMorel committed:
Add note about setSQLLogger()

BenMorel committed:
Don't catch Exception anymore

morozov committed:
When rendering SQL, only render the alias if it's different from the table name

Majkl578 committed:
Drop support for PostgreSQL <9.3

Majkl578 committed:
Drop support for SQL Server <2008

Majkl578 committed:
Drop support for SQL Anywhere <16

Majkl578 committed:
Drop PDO DB2 driver

Majkl578 committed:
Remove MsSQLKeywords

Majkl578 committed:
Reorder keywords and type mappings alphabetically

morozov committed:
Fixed build failure on SQL Server (sqlsrv)

The regression was introduced by 2b8c40d267b4ac699ea878dcca07332e1697978c.

Fixes #3060.

morozov committed:
Fixed build failure on SQL Server (pdo_sqlsrv)

Use a wrapped PDO Statement instead of a custom class for `pdo_sqlsrv` since `PDOSqlsrv\Statement` doesn't extend `PDOStatement` anymore.

morozov committed:
[DBAL-3079] Reworked the usage of PDO in PDOConnection from inheritance to composition

morozov committed:
[DBAL-3079] Added type hints to query-related method parameters and return values

morozov committed:
[DBAL-3079] Updated upgrade documentation

Majkl578 committed:
Revert complex type specification in Connection

Commit:
Added string mapping in SqlitePlatform - used by make bundle
  - fix 'Unknown database type string requested, Doctrine\DBAL\Platforms\SqlitePlatform may not support it.' error

Ocramius committed:
Merge pull request #3119 from DeyV/develop

Added "string" mapping in SqlitePlatform

Majkl578 committed:
Merge pull request #3113 from Majkl578/revert-complex-type-spec

Revert complex type specification in Connection

morozov committed:
Removed the logic of using BLOB columns for BINARY type of fields

morozov committed:
Removed support for DB-generated UUIDs

morozov committed:
Merge pull request #3211 from morozov/issues/3167

Removed support for DB-generated UUIDs

morozov committed:
Fixed and suppressed issues found by PHPStan on develop

Ocramius committed:
Merge pull request #3238 from morozov/phpstan-develop

Fixed and suppressed issues found by PHPStan on develop

morozov committed:
Removed dbal:import CLI command

morozov committed:
Merge pull request #3241 from morozov/issues/3237

Removed dbal:import CLI command

morozov committed:
Made the OFFSET in LIMIT queries non-nullable

morozov committed:
Merge pull request #3248 from morozov/non-nullable-offset

Made the OFFSET in LIMIT queries non-nullable integer defaulting to 0

Majkl578 committed:
Remove Doctrine\DBAL\Types\Type::getDefaultLength()

Majkl578 committed:
Remove Doctrine\DBAL\Types\Type::__toString()

Majkl578 committed:
Enable PHPStan in tests

Majkl578 commented on tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php:
```diff
@@ -743,11 +743,11 @@ public function testFetchAllSupportFetchClass()
 
         self::assertCount(1, $results);
-        self::assertInstanceOf(__NAMESPACE__.'\\MyFetchClass', $results[0]);
+        self::assertInstanceOf(MyFetchClass::class, $results[0]);
```
Not strictly needed, but affected by https://github.com/phpstan/phpstan-phpunit/issues/27.

morozov commented on composer.json:
```diff
@@ -21,6 +21,7 @@
         "doctrine/coding-standard": "^4.0",
         "jetbrains/phpstorm-stubs": "^2018.1.2",
         "phpstan/phpstan": "^0.10.1",
+        "phpstan/phpstan-phpunit": "dev-GetMockBuilderDynamicReturnTypeExtension-with-undefined-class as 0.10.0.999",
```
Note to self: make sure this patch is released first.

morozov commented on tests/Doctrine/Tests/DBAL/Driver/PDOExceptionTest.php:
```diff
@@ -24,7 +24,7 @@ class PDOExceptionTest extends DbalTestCase
      * The wrapped PDO exception mock.
      *
-     * @var \PDOException|\PHPUnit_Framework_MockObject_MockObject
+     * @var \PDOException
```
Yay!

morozov commented on tests/Doctrine/Tests/DBAL/Functional/Driver/IBMDB2/DB2StatementTest.php:
```diff
@@ -27,6 +29,7 @@ protected function setUp()
     public function testExecutionErrorsAreNotSuppressed()
     {
         $stmt = $this->_conn->prepare('SELECT * FROM SYSIBM.SYSDUMMY1 WHERE \'foo\' = ?');
+        assert($stmt instanceof Statement);
```
What else could it be?

morozov commented on tests/Doctrine/Tests/DBAL/Schema/MySqlSchemaManagerTest.php:
```diff
@@ -15,7 +16,7 @@ class MySqlSchemaManagerTest extends \PHPUnit\Framework\TestCase
     private $manager;
 
-    /** @var Connection */
+    /** @var Connection|MockObject */
```
It looks like the opposite of https://github.com/doctrine/dbal/pull/3260/files#diff-90748caec80a98aa0fcfef640916661bL27.

morozov replied:
I see. Makes sense.

I thought, using `phpstan-phpunit` would eliminate the necessity to explicitly annotate mocks as such. Looks like it's only the case for the return values of methods like `createMock()` which automatically get the traits of the object under test and the mock. But for properties, you still have to specify the type explictly. Does it sound about right?

ondrejmirtes replied:
Yes, it converts them in this case, currently hardcoded in PHPStan, but will be cleaner and part of the phpunit extension in 0.11.

morozov commented on tests/Doctrine/Tests/DBAL/Schema/SchemaTest.php:
```diff
@@ -449,6 +451,9 @@ public function testVisitsNamespaceVisitor()
+        $schema->visit($visitor);
+
+        self::doesNotPerformAssertions(); // FIXME
+
```
Redundant empty line.

morozov commented on tests/Doctrine/Tests/DBAL/Types/DateImmutableTypeTest.php:
```diff
@@ -24,7 +25,7 @@ class DateImmutableTypeTest extends \PHPUnit\Framework\TestCase
     {
         $this->type = Type::getType('date_immutable');
-        $this->platform = $this->prophesize(AbstractPlatform::class);
+        $this->platform = $this->getMockBuilder(AbstractPlatform::class)->getMock();
```
Probably could use `$this->createMock()` instead.

Majkl578 replied:
Yes, forgot to change this file. :smiley:

morozov commented on tests/Doctrine/Tests/Mocks/TaskMock.php:
```diff
@@ -1,84 +0,0 @@
- * TaskMock used for testing the CLI interface.
- * @author Nils Adermann <naderman@naderman.de>
- */
-class TaskMock extends \Doctrine\Common\Cli\Tasks\AbstractTask
```
:heart:

morozov committed:
Got rid of func_num_args in AbstractPlatform::getDummySelectSQL()

morozov committed:
Handle binding errors in OCI8Statement::execute() and MySQLiStatement::execute()

morozov committed:
Statement::fetchColumn() will throw an exception in the case of invalid index

morozov committed:
Moved rowCount() from Statement to ResultStatement

Majkl578 committed:
Enable strict types

morozov committed:
Merge pull request #3417 from morozov/issues/3358

Statement::fetchColumn() will throw an exception in the case of invalid index

morozov committed:
Transaction-related Statement methods return void

morozov committed:
Connection::setAutoCommit() will return void

Ocramius committed:
Merge pull request #3480 from morozov/transaction-void

Transaction-related Statement methods return void

morozov committed:
Converted Connection and Statement methods which returned false in case of a failure into void

morozov committed:
Merge pull request #3483 from morozov/row-count

Moved rowCount() from Statement to ResultStatement

morozov committed:
Merge pull request #3486 from morozov/conn-stmt-void

Converted Connection and Statement methods which returned false in case of a failure into void

morozov committed:
Connection::quote() can only quote strings

morozov committed:
Assert that the username is specified on the connection when listing DB2 tables

morozov committed:
Make the $tableName argument of _getPortableTableIndexesList() required

1. The argument is always available since the method is only called from listTableIndexes() which requires a table name.
2. The argument itself seems a workaround and only needed to bypass the fact that the SQLiteSchemaManager violates the method contract: instead of reformatting the provided data it fetches more data from the DB schema which requires a table name. This argument should be dropped completely later.

Ocramius committed:
Merge pull request #3488 from morozov/quote-only-string

Connection::quote() can only quote strings

morozov committed:
Changed the type of `$char` in `AbstractPlatform::getTrimExpression()` from `string|false` to `?string`

Ocramius committed:
Merge pull request #3491 from morozov/get-trim-expr

Changed the type of $char in AbstractPlatform::getTrimExpression()`from string|false to ?string

morozov committed:
Modified AbstractPlatform::getLocateExpression() and ::getSubstringExpression() signatures

Ocramius committed:
Merge pull request #3494 from morozov/get-locate-expr

Modified AbstractPlatform::getLocateExpression() and ::getSubstringExpression() signatures

morozov committed:
Reworked AbstractPlatform::get*Expression() methods

Ocramius committed:
Merge pull request #3498 from morozov/get-other-expr

Reworked AbstractPlatform::get*Expression() methods

morozov committed:
Reworked `AbstractSchemaManager::extractDoctrineTypeFromComment()` removed `::removeDoctrineTypeFromComment()`

morozov committed:
Removed DB2SchemaManager::_getPortableForeignKeyRuleDef()

morozov committed:
Merge pull request #2854 from Majkl578/strict-types

Enable strict types

Majkl578 committed:
Merge pull request #3502 from morozov/db2-get-portable-foreign-key-rule-def

Removed DB2SchemaManager::_getPortableForeignKeyRuleDef()

morozov committed:
Reworking driver exceptions

Ocramius committed:
Merge pull request #3505 from morozov/driver-exceptions

Reworked driver exceptions

morozov committed:
Removed errorCode() and errorInfo() methods from Connection and Statement APIs

Ocramius committed:
Merge pull request #3507 from morozov/error-code-info

Removed errorCode() and errorInfo() methods from Connection and Statement APIs

morozov committed:
Merge pull request #3501 from morozov/extract-type-from-comment

Reworked AbstractSchemaManager::extractDoctrineTypeFromComment(), removed ::removeDoctrineTypeFromComment()

Majkl578 commented:
Rebased, green on level 3, 65 more errors on level 7.

morozov commented:
> Rebased on top of develop, green on level 3, 65 more errors on level 7.

We can wrap it up at level 3 and continue to improve later in order to avoid future conflicts and new issues on Level 3 and lower. BTW, why `develop`? Is it possible to move it to `master`, same as `lib/`, especially given that we should have more freedom in making changes in tests?

Majkl578 commented:
> BTW, why develop?

Picked wrong target when rebasing and didn't want to start again on top of master. :|

morozov commented:
> Picked wrong target when rebasing and didn't want to start again on top of master. :|

Could you at least roll it back? I would like to have it done in `master`, and backporting to `master` and rebasing `develop` on top of it would be unnecessary extra work.

morozov commented:
Closing in favor of #3582. Thanks, @Majkl578.
